README.md
# Go Pretty Time

[![Go Report Card](https://goreportcard.com/badge/gojp/goreportcard)](https://goreportcard.com/report/github.com/andanhm/go-prettytime)
[![Build Status](https://travis-ci.org/andanhm/go-prettytime.svg?branch=master)](https://travis-ci.org/andanhm/go-prettytime)
[![GoDoc](https://camo.githubusercontent.com/3de3bba30c9355c0d919804e7b31e6b504af74e2/68747470733a2f2f676f646f632e6f72672f6769746875622e636f6d2f6e617468616e792f6c6f6f7065723f7374617475732e737667)](https://godoc.org/github.com/andanhm/go-prettytime)
[![CodeCov](https://codecov.io/gh/icza/minquery/branch/master/graph/badge.svg)](https://codecov.io/gh/andanhm/go-prettytime)

Format Go date time in a _`pretty`_ way. ex : just now, a minute ago, 2 hours ago , 3 minutes ago

Inspired by the `John Resig` Pretty Date [plug-in] for JQuery

```
prettytime.Format("2008-01-28T20:24:17Z") // => "2 hours ago"
prettytime.Format("2008-01-27T22:24:17Z") // => "Yesterday"
prettytime.Format("2008-01-26T22:24:17Z") // => "2 days ago"
prettytime.Format("2008-01-14T22:24:17Z") // => "2 weeks ago"
```

## Install

```bash
go get github.com/andanhm/go-prettytime
```

### Example

```go
package main

import (
	"log"
	"time"

	"github.com/andanhm/go-prettytime"
)

const (
	layout = "2006-01-02T15:04:05Z"
)

func main() {
	t, err := time.Parse(layout, "2008-01-28T20:24:17Z")
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("%s \n", prettytime.Format(t))
}
// Output: 13 years ago
```

### Contributions

Feel free to fork and add features, fix bugs and your pull request is more than welcome ❤

[go-prettytime]: https://pkg.go.dev/github.com/andanhm/go-prettytime
[plug-in]: http://ejohn.org/blog/javascript-pretty-date/


CHANGELOG.md
# 1.1.0

-   Human readable format
-   Go Documentation
-   Travis integration with unit testing and code coverage

# 1.0.0

-   Human readable format
-   Go Documentation
-   Travis integration with unit testing and code coverage


time_test.go
package prettytime_test

import (
	"fmt"
	"testing"
	"time"

	. "github.com/andanhm/go-prettytime"
)

func TestFormat(t *testing.T) {
	tests := []struct {
		name string
		t    time.Time
		want string
	}{
		{name: "Just now", t: time.Now(), want: "just now"},
		{name: "Second", t: time.Now().Add(
			time.Hour*time.Duration(0) +
				time.Minute*time.Duration(0) +
				time.Second*time.Duration(1)),
			want: "1 second from now",
		},
		{name: "SecondAgo", t: time.Now().Add(
			time.Hour*time.Duration(0) +
				time.Minute*time.Duration(0) +
				time.Second*time.Duration(-1)),
			want: "1 second ago"},
		{name: "Minutes", t: time.Now().Add(time.Hour*time.Duration(0) +
			time.Minute*time.Duration(59) +
			time.Second*time.Duration(59)), want: "60 minutes from now"},
		{name: "Tomorrow", t: time.Now().AddDate(0, 0, 1), want: "tomorrow"},
		{name: "Yesterday", t: time.Now().AddDate(0, 0, -1), want: "yesterday"},
		{name: "Week", t: time.Now().AddDate(0, 0, 7), want: "1 week from now"},
		{name: "WeekAgo", t: time.Now().AddDate(0, 0, -7), want: "1 week ago"},
		{name: "Month", t: time.Now().AddDate(0, 1, 0), want: "1 month from now"},
		{name: "MonthAgo", t: time.Now().AddDate(0, -1, 0), want: "1 month ago"},
		{name: "Year", t: time.Now().AddDate(50, 0, 0), want: "50 years from now"},
		{name: "YearAgo", t: time.Now().AddDate(-2, 0, 0), want: "2 years ago"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if gotTimeSince := Format(tt.t); gotTimeSince != tt.want {
				t.Errorf("Format() = %v, want %v", gotTimeSince, tt.want)
			}
		})
	}
}

func ExampleFormat() {
	timeSlots := []struct {
		name string
		t    time.Time
	}{
		{name: "Just now", t: time.Now()},
		{name: "Second", t: time.Now().Add(
			time.Hour*time.Duration(0) +
				time.Minute*time.Duration(0) +
				time.Second*time.Duration(1)),
		},
		{name: "SecondAgo", t: time.Now().Add(
			time.Hour*time.Duration(0) +
				time.Minute*time.Duration(0) +
				time.Second*time.Duration(-1)),
		},
		{name: "Minutes", t: time.Now().Add(time.Hour*time.Duration(0) +
			time.Minute*time.Duration(59) +
			time.Second*time.Duration(59))},
		{name: "Tomorrow", t: time.Now().AddDate(0, 0, 1)},
		{name: "Yesterday", t: time.Now().AddDate(0, 0, -1)},
		{name: "Week", t: time.Now().AddDate(0, 0, 7)},
		{name: "WeekAgo", t: time.Now().AddDate(0, 0, -7)},
		{name: "Month", t: time.Now().AddDate(0, 1, 0)},
		{name: "MonthAgo", t: time.Now().AddDate(0, -1, 0)},
		{name: "Year", t: time.Now().AddDate(2, 0, 0)},
		{name: "YearAgo", t: time.Now().AddDate(-2, 0, 0)},
	}

	for _, timeSlot := range timeSlots {
		fmt.Printf("%s = %v\n", timeSlot.name, Format(timeSlot.t))
	}
}

func TestFormatYear(t *testing.T) {
	now := time.Now()

	oneYearFromNow := now.AddDate(1, 0, 0)
	gotTimeSince := Format(oneYearFromNow)
	expected := "1 year from now"
	if gotTimeSince != expected {
		t.Errorf("got %v, want %v", expected, gotTimeSince)
	}
}


_example/main.go
package main

import (
	"fmt"
	"time"

	prettyTime "go-prettytime"
)

var now = time.Now()
var timeFormat = time.RFC3339
var interval  = time.Duration(10)

func main() {
	fmt.Println(now.Format(timeFormat), "->", prettyTime.Format(now))

	Time := now.Add(time.Hour*time.Duration(0) + time.Minute*time.Duration(0) + time.Second*5)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.Add(time.Hour*time.Duration(0) + time.Minute*time.Duration(3) + time.Second)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, 0, 1)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, 0, -1)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, 0, 7)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, 0, -7)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, 1, 0)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	Time = now.AddDate(0, -1, 0)
	fmt.Println(Time.Format(timeFormat), "->", prettyTime.Format(Time))

	go func() {
		for range time.Tick(time.Second * interval) {
			now.Add(time.Hour*time.Duration(0) + time.Minute*time.Duration(0) + time.Second*interval)
		}
	}()
	time.Sleep(time.Second * interval)
	for range time.Tick(time.Second * interval) {
		fmt.Println(time.Now().Format(timeFormat), "->",prettyTime.Format(now))
	}
}


time.go
//Package prettytime provides utility functions for calculating and presenting
//in human readable and understandable form.
package prettytime

import (
	"strconv"
	"strings"
	"time"
)

// Unix epoch (or Unix time or POSIX time or Unix timestamp)  1 year (365.24 days)
const infinity float64 = 31556926 * 1000

// Handler function which determines the time difference based on defined time spams
func handler(timeIntervalThreshold float64, timePeriod, message string) func(float64) string {
	return func(difference float64) string {
		var str strings.Builder
		n := difference / timeIntervalThreshold
		nStr := strconv.FormatFloat(n, 'f', 0, 64)
		str.WriteString(nStr)
		str.WriteString(" ")
		str.WriteString(timePeriod)
		if int(n) > 1 {
			str.WriteString("s ")
			str.WriteString(message)
			return str.String()
		}
		str.WriteString(" ")
		str.WriteString(message)
		return str.String()
	}
}

// timeLapse condition struct
type timeLapse struct {
	// Time stamp threshold to handle the time lap condition
	Threshold float64
	// Handler function which determines the time lapse based on the condition
	Handler func(float64) string
}

var timeLapses = []timeLapse{
	{Threshold: -31535999, Handler: handler(-31536000, "year", "from now")},
	{Threshold: -2591999, Handler: handler(-2592000, "month", "from now")},
	{Threshold: -604799, Handler: handler(-604800, "week", "from now")},
	{Threshold: -172799, Handler: handler(-86400, "day", "from now")},
	{Threshold: -86399, Handler: func(diff float64) string {
		return "tomorrow"
	}},
	{Threshold: -3599, Handler: handler(-3600, "hour", "from now")},
	{Threshold: -59, Handler: handler(-60, "minute", "from now")},
	{Threshold: -0.9999, Handler: handler(-1, "second", "from now")},
	{Threshold: 1, Handler: func(diff float64) string {
		return "just now"
	}},
	{Threshold: 60, Handler: handler(1, "second", "ago")},
	{Threshold: 3600, Handler: handler(60, "minute", "ago")},
	{Threshold: 86400, Handler: handler(3600, "hour", "ago")},
	{Threshold: 172800, Handler: func(diff float64) string {
		return "yesterday"
	}},
	{Threshold: 604800, Handler: handler(86400, "day", "ago")},
	{Threshold: 2592000, Handler: handler(604800, "week", "ago")},
	{Threshold: 31536000, Handler: handler(2592000, "month", "ago")},
	{Threshold: infinity, Handler: handler(31536000, "year", "ago")},
}

//Format returns a string describing how long it has been since
//the time argument passed int
func Format(t time.Time) (timeSince string) {
	timestamp := t.Unix()
	now := time.Now().Unix()

	if timestamp > now || timestamp <= 0 {
		timeSince = ""
	}

	timeElapsed := float64(now - timestamp)
	for _, formatter := range timeLapses {
		if timeElapsed < formatter.Threshold {
			timeSince = formatter.Handler(timeElapsed)
			break
		}
	}
	return timeSince
}
